sw/device/lib/base/memory_perftest.c

```diff
@@ -0,0 +1,308 @@
+  CHECK(start_cycles < end_cycles);
+  return end_cycles - start_cycles;
+}
+
+static inline uint64_t read_cycle_counter() {
+#ifdef OT_PLATFORM_RV32
+  // We cannot read the entire 64-bit mcycle counter at once on RISCV32.
+  uint32_t mcycleh, mcycleh2, mcycle_lower;
+  do {
+    CSR_READ(CSR_REG_MCYCLEH, &mcycleh);
+    CSR_READ(CSR_REG_MCYCLE, &mcycle_lower);
+    CSR_READ(CSR_REG_MCYCLEH, &mcycleh2);
+  } while (mcycleh != mcycleh2);
+
+  return ((uint64_t)mcycleh << 32) | mcycle_lower;
+#else
+  static_assert(false, "Not implemented");
+#endif
+}
+
+static inline bool compute_statistics(const uint64_t *durations,
+                                      size_t num_durations, uint32_t *out_mean,
+                                      uint32_t *out_variance) {
+  uint64_t duration_sum = 0;
+  for (size_t rep = 0; rep < num_durations; ++rep) {
+    if (duration_sum > UINT64_MAX - durations[rep]) {
+      LOG_WARNING("Summing durations would overflow.");
+      return false;
+    }
+    duration_sum += durations[rep];
+  }
+  CHECK(num_durations > 0);
+  const uint64_t duration_mean =
+      udiv64_slow(duration_sum, num_durations, /*rem_out=*/NULL);
+
+  uint64_t sum_squared_differences = 0;
+  for (size_t rep = 0; rep < num_durations; ++rep) {
+    uint64_t difference;
+    if (durations[rep] > duration_mean) {
+      difference = durations[rep] - duration_mean;
+    } else if (durations[rep] < duration_mean) {
+      difference = duration_mean - durations[rep];
+    } else {
+      continue;
+    }
+    if (difference > udiv64_slow(UINT64_MAX, difference, /*rem_out=*/NULL)) {
+      LOG_WARNING("Squaring difference would overflow.");
+      return false;
+    }
+    if (sum_squared_differences > UINT64_MAX - difference * difference) {
+      LOG_WARNING("Adding squared difference to sum would overflow.");
+      return false;
+    }
+    sum_squared_differences += difference * difference;
+  }
+
+  CHECK(num_durations > 1);
+  const uint64_t duration_variance = udiv64_slow(
+      sum_squared_differences, (uint64_t)num_durations - 1, /*rem_out=*/NULL);
+
+  if (duration_mean > UINT32_MAX) {
+    LOG_WARNING("Computed mean cannot fit in 32 bits");
+    return false;
+  }
+  if (duration_variance > UINT32_MAX) {
+    LOG_WARNING("Computed variance cannot fit in 32 bits");
+    return false;
+  }
+  *out_mean = (uint32_t)duration_mean;
+  *out_variance = (uint32_t)duration_variance;
+  return true;
+}
+
+static inline void fill_buf_arbitrary_values(uint8_t *buf, size_t len) {
+  // TODO(dmcardle) Use an algorithm that maximizes the sequence's Hamming
+  // distance? It's entirely possible that this code generates sequences that
+  // hit a fixed point and repeat forever; I haven't really analyzed this. At
+  // minimum, we should make sure it doesn't get stuck in a small cycle before
+  // generating PERFTEST_MAX_NUM_REPS * PERFTEST_BUF_LEN bytes.
+  static uint32_t seed = 0xbeefbeef;
+  seed *= 13;
```

Comment:
Have a look at `sw/device/testing/rand_testutils.c`.

I'd say we should enhance those routines to allow seeding and  use them to generate the numbers used for this sequence.

Reply:
Aha! I was looking for an RNG!